go/tools/builders/nogo_change.go

```diff
@@ -0,0 +1,350 @@
+					continue
+				}
+				if start > end {
+					allErrors = append(allErrors, fmt.Errorf("invalid fix: pos %v > end %v", start, end))
+					continue
+				}
+				if eof := token.Pos(file.Base() + file.Size()); end > eof {
+					allErrors = append(allErrors, fmt.Errorf("invalid fix: end %v past end of file %v", end, eof))
+					continue
+				}
+
+				edit := Edit{Start: file.Offset(start), End: file.Offset(end), New: string(edit.NewText)}
+				fileRelativePath, err := filepath.Rel(cwd, file.Name())
+				if err != nil {
+					fileRelativePath = file.Name() // fallback logic
+				}
+				c.AddEdit(fileRelativePath, analyzer, edit)
+			}
+		}
+	}
+
+	if len(allErrors) > 0 {
+		return c, fmt.Errorf("errors: %v", allErrors)
+	}
+	return c, nil
+}
+
+
+// AddEdit adds an edit to the Change, organizing by file and analyzer.
+func (c *Change) AddEdit(file string, analyzer string, edit Edit) {
+	// Ensure the FileEdits structure exists for the file
+	fileEdits, exists := c.FileToEdits[file]
+	if !exists {
+		fileEdits = FileEdits{
+			AnalyzerToEdits: make(map[string][]Edit),
+		}
+		c.FileToEdits[file] = fileEdits
+	}
+
+	// Append the edit to the list of edits for the analyzer
+	fileEdits.AnalyzerToEdits[analyzer] = append(fileEdits.AnalyzerToEdits[analyzer], edit)
+}
+
+
+
+// Flatten merges all edits for a file from different analyzers into a single map of file-to-edits.
+// Edits from each analyzer are processed in a deterministic order, and overlapping edits are skipped.
+func Flatten(change Change) map[string][]Edit {
+	fileToEdits := make(map[string][]Edit)
+
+	for file, fileEdits := range change.FileToEdits {
+		// Get a sorted list of analyzers for deterministic processing order
+		analyzers := make([]string, 0, len(fileEdits.AnalyzerToEdits))
+		for analyzer := range fileEdits.AnalyzerToEdits {
+			analyzers = append(analyzers, analyzer)
+		}
+		sort.Strings(analyzers)
+
+		mergedEdits := make([]Edit, 0)
+
+		for _, analyzer := range analyzers {
+			edits := fileEdits.AnalyzerToEdits[analyzer]
+
+			// Deduplicate and sort edits for the current analyzer
+			edits, _ = UniqueEdits(edits)
+
+			// Merge edits into the current list, checking for overlaps
+			candidateEdits := append(mergedEdits, edits...)
+			candidateEdits, invalidIndex := UniqueEdits(candidateEdits)
+			if invalidIndex >= 0 {
+				// Skip edits from this analyzer if merging them would cause overlaps.
+				// Apply the non-overlapping edits first. After that, a rerun of bazel build will
+				// allows these skipped edits to be applied separately.
+				// Note the resolution happens to each file independently.
+				continue
+			}
+
+			// Update the merged edits
+			mergedEdits = candidateEdits
+		}
+
+		// Store the final merged edits for the file
+		fileToEdits[file] = mergedEdits
+	}
+
+	return fileToEdits
+}
+
+
+// ToCombinedPatch converts all edits to a single consolidated patch.
+func ToCombinedPatch(fileToEdits map[string][]Edit) (string, error) {
```

Comment:
done. 

I tried to make structs types private too, but that make the type name like variables such as 'edit', 'fileEdits', so I still keep the uppercase for each, to avoid naming collision or accidental use, I prefix them with Nogo.